File: src/tree/lc_144_preorder_traversal.cc
#include "tree/lc_144_preorder_traversal.h"

#include "gtest/gtest.h"

namespace algo {
namespace tree {
class PreorderTraversalTest : public ::testing::Test {
 protected:
  void SetUp() override { exec = new PreorderTraversal(); }

  void TearDown() override { delete exec; }

 protected:
  PreorderTraversal* exec{nullptr};
};

TEST_F(PreorderTraversalTest, preorderTraversal1) {
  std::vector<int> nums1 = {1, -1, 2, 3, -1, -1, -1};
  TreeNode<int>* root1 = ToBinaryTree(nums1);
  std::vector<int> res1 = exec->preorderTraversal1(root1);
  ASSERT_EQ(1, res1.at(0));
  ASSERT_EQ(2, res1.at(1));
  ASSERT_EQ(3, res1.at(2));
  FreeBinaryTree(root1);

  std::vector<int> nums2 = {};
  TreeNode<int>* root2 = ToBinaryTree(nums2);
  std::vector<int> res2 = exec->preorderTraversal1(root2);
  ASSERT_EQ(0, res2.size());
  FreeBinaryTree(root2);

  std::vector<int> nums3 = {1, -1, -1};
  TreeNode<int>* root3 = ToBinaryTree(nums3);
  std::vector<int> res3 = exec->preorderTraversal1(root3);
  ASSERT_EQ(1, res3.at(0));
  FreeBinaryTree(root3);
}

TEST_F(PreorderTraversalTest, preorderTraversal2) {
  std::vector<int> nums1 = {1, -1, 2, 3, -1, -1, -1};
  TreeNode<int>* root1 = ToBinaryTree(nums1);
  std::vector<int> res1 = exec->preorderTraversal2(root1);
  ASSERT_EQ(1, res1.at(0));
  ASSERT_EQ(2, res1.at(1));
  ASSERT_EQ(3, res1.at(2));
  FreeBinaryTree(root1);

  std::vector<int> nums2 = {};
  TreeNode<int>* root2 = ToBinaryTree(nums2);
  std::vector<int> res2 = exec->preorderTraversal2(root2);
  ASSERT_EQ(0, res2.size());
  FreeBinaryTree(root2);

  std::vector<int> nums3 = {1, -1, -1};
  TreeNode<int>* root3 = ToBinaryTree(nums3);
  std::vector<int> res3 = exec->preorderTraversal2(root3);
  ASSERT_EQ(1, res3.at(0));
  FreeBinaryTree(root3);
}
}  // namespace tree
}  // namespace algo

File: src/tree/lc_144_preorder_traversal.go
package tree

func preorderTraversal1(root *TreeNode) (vals []int) {
	var preorder func(node *TreeNode)
	preorder = func(node *TreeNode) {
		if node == nil {
			return
		}

		vals = append(vals, node.Val)
		preorder(node.Left)
		preorder(node.Right)
	}

	preorder(root)
	return
}

func preorderTraversal2(root *TreeNode) (vals []int) {
	if root == nil {
		return vals
	}

	var stack []*TreeNode

	stack = append(stack, root)

	for len(stack) != 0 {
		node := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		vals = append(vals, node.Val)

		if node.Right != nil {
			stack = append(stack, node.Right)
		}

		if node.Left != nil {
			stack = append(stack, node.Left)
		}
	}

	return
}
